[export] add SchemaCheckMode testing for pre-dispatch export, OpInfo

pianpwk opened the pull request:
This adds a new dispatch mode, PreDispatchSchemaCheckMode, built on top of SchemaCheckMode, used for verifying op schemas for functionalization for PreDispatch IR. More specifically, the mode runs in eager mode on concrete inputs, checking if op schemas incorrectly claim to be functional, but are aliasing or mutating. This mode is pushed to the pre-dispatch mode stack, and run before decompositions. 

Current testing is hooked up to OpInfo, containing 1103 tests on 600 unique ops. Below is a list of ops that fail testing. One caveat is we only raise errors on ops that claim to be functional - if an op schema admits aliasing or mutating but fails testing for the other, it still may decompose further and become functional.

List of failed ops:
```
aten.atleast_1d.default
aten.atleast_2d.default
aten.atleast_3d.default
aten.cartesian_prod.default
aten.conj_physical.default
aten.alpha_dropout.default
aten.feature_dropout.default
aten.feature_alpha_dropout.default
aten.unsafe_chunk.default
```

bdhirsh commented on test/export/opinfo_schema.py:
```diff
@@ -0,0 +1,109 @@
+test_classes = {}
+
+
+class PreDispatchSchemaCheckMode(SchemaCheckMode):
```
Hmm, is the implementation different enough to warrant a new class? Another option is just to use the existing `SchemaCheckMode`, and give it a new `bool is_predispatch` argument, that you use to optionally set the `_dispatch_key` field on it.

pianpwk replied:
I think the main reason is to have access to the aten op to check if it's aliasing or mutating before we raise an error? Without subclassing I'm not sure how to access the op and check from the OpInfo object.

pianpwk committed:
[export] add SchemaCheckMode testing for pre-dispatch export, OpInfo (#125481)

Summary:
This adds a new dispatch mode, PreDispatchSchemaCheckMode, built on top of SchemaCheckMode, used for verifying op schemas for functionalization for PreDispatch IR. More specifically, the mode runs in eager mode on concrete inputs, checking if op schemas incorrectly claim to be functional, but are aliasing or mutating. This mode is pushed to the pre-dispatch mode stack, and run before decompositions.

Current testing is hooked up to OpInfo, containing 1103 tests on 600 unique ops. Below is a list of ops that fail testing. One caveat is we only raise errors on ops that claim to be functional - if an op schema admits aliasing or mutating but fails testing for the other, it still may decompose further and become functional.

List of failed ops:
```
aten.atleast_1d.default
aten.atleast_2d.default
aten.atleast_3d.default
aten.cartesian_prod.default
aten.conj_physical.default
aten.alpha_dropout.default
aten.feature_dropout.default
aten.feature_alpha_dropout.default
aten.unsafe_chunk.default
```


Reviewed By: tugsbayasgalan

Differential Revision: D56943527

Pulled By: pianpwk
